app/v2/controllers/V2MovementsController.scala

```diff
@@ -228,20 +240,75 @@ class V2MovementsControllerImpl @Inject() (
     }
 
   def getMessage(movementType: MovementType, movementId: MovementId, messageId: MessageId): Action[AnyContent] =
-    (authActionNewEnrolmentOnly andThen acceptHeaderActionProvider(acceptOnlyJson = false)).async {
+    (authActionNewEnrolmentOnly andThen acceptHeaderActionProvider(jsonAndJsonWrappedXmlAcceptHeaders)).async {
+      implicit request =>
+        implicit val hc: HeaderCarrier = HeaderCarrierConverter.fromRequest(request)
+        (for {
+          messageSummary <- persistenceService.getMessage(request.eoriNumber, movementType, movementId, messageId).asPresentation
+          acceptHeader = request.headers.get(HeaderNames.ACCEPT).get
+          body <-
+            if (messageSummary.uri.isDefined) processLargeMessage(movementId, movementType, messageSummary, acceptHeader)
+            else processSmallMessage(movementId, movementType, messageSummary, acceptHeader)
+        } yield body).fold(
+          presentationError => Status(presentationError.code.statusCode)(Json.toJson(presentationError)),
+          response => Ok.chunked(response)
+        )
+    }
+
+  private def processLargeMessage(movementId: MovementId, movementType: MovementType, messageSummary: MessageSummary, acceptHeader: String)(implicit
+    hc: HeaderCarrier
+  ): EitherT[Future, PresentationError, Source[ByteString, _]] =
+    acceptHeader match {
+      case VersionedRouting.VERSION_2_ACCEPT_HEADER_VALUE_JSON =>
+        EitherT.leftT[Future, Source[ByteString, _]](PresentationError.notAcceptableError("Large messages cannot be returned as json"))
+      case VersionedRouting.VERSION_2_ACCEPT_HEADER_VALUE_JSON_XML | VersionedRouting.VERSION_2_ACCEPT_HEADER_VALUE_JSON_XML_HYPHEN =>
+        for {
+          bodyStream <- objectStoreService.getMessage(messageSummary.uri.get).asPresentation
+          json: JsObject = Json.toJson(HateoasMovementMessageResponse(movementId, messageSummary.id, messageSummary, movementType)).as[JsObject]
+          stream <- mergeStreamIntoJson(json.fields, "body", bodyStream)
+        } yield stream
+    }
+
+  private def processSmallMessage(movementId: MovementId, movementType: MovementType, messageSummary: MessageSummary, acceptHeader: String)(implicit
+    hc: HeaderCarrier
+  ): EitherT[Future, PresentationError, Source[ByteString, _]] =
+    acceptHeader match {
+      case VersionedRouting.VERSION_2_ACCEPT_HEADER_VALUE_JSON =>
```

Comment:
Possibly combine with process smallMessage since they have the same match conditions.

Reply:
as discussed, same match conditions but very different implementations so would over complicate and make the function less readable in my opinion